### PR TITLE
Antigen presentation

### PR DESCRIPTION
**Pull request by timodonnell**
Infrastructure for antigen presentation models, which integrate MHC affinity, cleavage, and other features.

This is currently a WIP. My goal is to get ensemble training working and then merge this.

### CODE REVIEW

**Comment by timodonnell** [comment]
Ok, this is pretty much in place now.

File: mhcflurry/common.py [hunk 1]
```diff
@@ -127,3 +133,109 @@ def shuffle_split_list(indices, fraction=0.5):
...
+    print("Generated dataframe hash in %0.2f sec" % (time.time() - start))
+    return result
+
+
+def freeze_object(o):
+    """
+    Recursively convert nested dicts and lists into frozensets and tuples.
+    """
+    if isinstance(o, dict):
+        return frozenset({k: freeze_object(v) for k, v in o.items()}.items())
```
Comment (iskandr) [inline mhcflurry/common.py]: Want to freeze the keys too? Or are tuple keys never going to come up here?

File: mhcflurry/antigen_presentation/__init__.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from .presentation_model import PresentationModel
```
Comment (iskandr) [inline mhcflurry/antigen_presentation/__init__.py]: Is PresentationModel in this PR?

Comment (timodonnell) [inline mhcflurry/antigen_presentation/__init__.py]: It should be, is it not showing up for you?

File: mhcflurry/common.py (hunk 1 above)
Comment (timodonnell) [inline mhcflurry/common.py]: Not sure I understand -- anything that can be a key in a dict (e.g. a tuple) is already immutable, so there's no need to handle those

**Comment by iskandr** [comment]
Ahhh, you're right, sorry.

**Comment by timodonnell** [comment]
Cool if I merge this, @iskandr ?

**Comment by iskandr** [comment]
:+1:

**Pull request merged**